### ISSUE TITLE
Comment references support for VS Code documentation popup

### ISSUE THREAD

**Issue by Sominemo** [Dart-Code/Dart-Code#3648]
**Is your feature request related to a problem? Please describe.**
It's frustrating to manually search for an identifier in code when it's used in an entities' documentation as a comment reference.

**Describe the solution you'd like**
Documentation popups provided by Dart Code need to add support for active links when comment references are used.

**Describe alternatives you've considered**
Since comment references are supported in the editing window, I usually jump to the doc comment itself, and then click the comment reference from there.

**Additional context**
As [Effective Dart](https://dart.dev/guides/language/effective-dart/documentation#do-use-square-brackets-in-doc-comments-to-refer-to-in-scope-identifiers) states, you can use square brackets to refer to fields and methods, and this annotation is used by dartdoc to generate active links in documentation.

The references are already being highlighted correctly in the code itself with working "Go to Definition" etc. options, but not in the VS Code's documentation popup, these display plain text in square brackets instead.
![image](https://user-images.githubusercontent.com/19842935/139884603-43ab99a0-dd6a-4c59-9c68-974de33ebdc2.png)
![image](https://user-images.githubusercontent.com/19842935/139885210-e8cde24f-6c18-49fa-a732-15368a96b02a.png)

**Comment by DanTup** [comment]
This will require the markdown parser can provide is offsets/lengths for the references in a block:

https://github.com/dart-lang/markdown/issues/369

**Comment by sadespresso** [comment]
Do we have any update on this 😁

**Comment by DanTup** [comment]
No updates yet. This still requires changes to the Markdown package first.

**Comment by Number-3434** [comment]
Any updates yet?
Would it be possible to somehow use `dart doc` to generate a web page with references, and then show that in the hover instead?

**Comment by Number-3434** [comment]
Could you parse the reference as a inline code block (`like this`), so that we don't see the braces (`[]`)?

**Comment by Number-3434** [comment]
Any updates yet?

@srawlins This depends on [Provide offest/lengths for nodes](https://github.com/dart-lang/markdown/issues/369). Could this be fixed easily (and quickly)?

**Comment by srawlins** [comment]
> Could this be fixed easily (and quickly)?

No it would be a pretty big effort. Right now the markdown package does not even parse Markdown text into a syntax tree which represents the Markdown text, https://github.com/dart-lang/markdown/issues/364. It goes straight into an AST for the rendered HTML. :(

**Comment by DanTup** [comment]
This came up again in https://github.com/Dart-Code/Dart-Code/issues/4975 and I was thinking a bit more about it. I wonder whether offsets from the markdown parser is really what we need for these. It seems we already have the comment references in the AST (which we use for Semantic Tokens and Go-to-Definition), but for hovers we only have access to the Element model. I'm not sure that re-parsing the markdown content to get these references is the right approach.

@bwilkerson since in the hover we only have a single `Element` to look at, I wonder if we could fetch the AST for the unit that contains it, locate the dartdoc node for that element, and then `visit` it to find the comment references. It does mean fetching a resolved AST for a file that might not be open (and therefore not cached) in every hover request though and I don't know if that's great.

**Comment by bwilkerson** [comment]
If I'm understanding the request correctly, we want to insert a link in place of a comment reference, which will require not only knowing the range of text to replace but also the target of the link. I can't think of any existing way other than getting the resolved AST to get the target.

We could try adding this support to see what the performance implications are, but probably only if it could be done quickly (because I don't want to expend too much effort if we might not be able to use it. If it adds too large a delay in getting hover information there might be some optimizations we could consider (such as storing more information in the element model, or having a special "resolve this comment only" mode in the resolver), but those are both likely to be fairly large projects.

> ... since in the hover we only have a single `Element` to look at ...

It's possible that the doc comment for a given element might not be in the compilation unit in which the element is declared. For members that override an inherited member we allow the docs to be copied from an overridden member, which might cause us to need to look in multiple compilation units in order to find it. Adding a field to `Element` to record the file / element / etc. where the doc comment was found would ensure a single resolution request, but that's again more work.

**Comment by DanTup** [comment]
@Number-3434 

Actually, my idea was as Brian understood. The problem here isn't specifically knowing the range of the reference, but rather knowing what references there are so we know what target links we need to include (whether inline, or appended).

For example if the markdown contains `[MyClass]` then we'd need to a) know that it includes this reference and b) know where the declaration of `MyClass` is. That information will generally come together, so to have the target we'll probably also have the offset/length of the reference.

My comments above about this being complicated by not having the offset/length was a bit misguided. For some of the other requests (like adding semantic tokens for code blocks inside the comments) we probably do need to re-parse the markdown and know where those ranges are, but for these references we should instead use the existing data that's already been parsed.

@bwilkerson 

> It's possible that the doc comment for a given element might not be in the compilation unit in which the element is declared. For members that override an inherited member we allow the docs to be copied from an overridden member, which might cause us to need to look in multiple compilation units in order to find it.

You're right - although I think this is already handled because we already do this traversal to find the string documentation today (we compute the `documentedElement` by looking through the related elements to find one with non-null docs). So I think once we've done that, we could use the final element as the one to fetch a resolved AST for.

Something to try out I guess, although I don't know if it's something I'll get to in the immediate future.

**Comment by bwilkerson** [comment]
> ... we could use the final element as the one to fetch a resolved AST for.

Good point. That simplifies the performance question by making it an O(1) operation.

> Something to try out I guess ...

I think a good first step would be to compute the `ResolvedUnitResult` for the unit containing the doc comment, without doing anything with the result, and see what kind of performance impact that would have. Not sure how much effort that would require, and definitely not our highest priority task.

**Comment by Number-3434** [comment]
Just out of curiosity, once you've found the `[references]` in the dartdoc, how hard is it to get the actual reference?

**Comment by DanTup** [comment]
@Number-3434 that probably depends on how we found them. If we got them via the AST, we should already have a route to that info. If we discover them by parsing the markdown, we'd need to resolve the identifiers based on the scope. I don't know how difficult this would be (because I'm not familiar with the code that normally does this when parsing the docs), but it feels like duplicating work that's already done (and may be something else place to update if a feature like https://github.com/dart-lang/sdk/issues/50702 is implemented).

I think it would make more sense to try out the ideas above before looking at other ways to do this.

**Comment by Number-3434** [comment]
Maybe there could be a temporary setting for this (e.g. `dart.commentReferences`), which defaults to <kbd>"off"</kbd>, and has a value of <kbd>"openFiles"</kbd>, which will make comment references be parsed only if there is already and AST for it?

**Comment by bwilkerson** [comment]
> ... it feels like duplicating work that's already done (and may be something else place to update if a feature like dart-lang/sdk#50702 is implemented).

It would definitely duplicate existing code. I don't know how hard it would be to rewrite the code to be sharable, but I agree that it's better to try the more straightforward implementation first and consider implementations like this if the simple path isn't performant enough.

If you're interested, we are actively implementing the doc import proposal. Unless we find some technical issues that prevent it I'm fairly confident that that's going to happen. Which is a good thing because then the analyzer and dartdoc will have the same semantics for resolving these references, allowing the analyzer to have better support for dartdoc in the future.

> Btw, roughly how long does it take to create a AST?

I don't have those numbers at hand (and don't remember), but it is, of course, dependent on the length of the source file being parsed and/or resolved.

> Maybe there could be a temporary setting for this ...

There could be, but my guess is that most users want hovers primarily for code that's outside the libraries currently in open editors, so I'm not sure how much value that would add. I might be wrong about that, of course.
